components/launcher/src/server/mod.rs

```diff
@@ -95,29 +102,6 @@ impl Server {
         Ok((channel, supervisor, pipe))
     }
 
-    fn remove_pipe(&self) {
-        if fs::remove_file(&self.pipe).is_err() {
-            error!("Could not remove old pipe to supervisor {}", self.pipe);
-        } else {
-            debug!("Removed old pipe to supervisor {}", self.pipe);
-        }
-    }
-
-    #[allow(unused_must_use)]
-    fn reload(&mut self) -> Result<()> {
-        self.supervisor.kill();
-        self.supervisor.wait();
```

Comment:
How is this handled in the new code... I'm missing it.

Reply:
The server is now scoped within the `loop` on line 397, so when it restarts the old server is dropped (triggering the cleanup) and a new server is created.

Reply:
I don't think that's true... `supervisor` is an instance of [std::process::Child](https://doc.rust-lang.org/1.27.2/std/process/struct.Child.html), and the docs say:

> There is no implementation of Drop for child processes, so if you do not ensure the Child has exited then it will continue to run, even after the Child handle to the child process has gone out of scope.

Reply:
I'm not relying on the drop to terminate `supervisor: Child`. On the contrary, the outer loop is only ever restarted when `server.tick()` returns `Err`, which implies the supervisor has already exited.

The tests demonstrate this works and it doesn't leave an old supervisor process running:

```
➤ env RUST_LOG=1 HAB_FEAT_TEST_EXIT=/tmp/exit-file TESTING_FS_ROOT=(mktemp -d)  HAB_SUP_BINARY=(pwd)/target/debug/hab-sup HAB_LAUNCH_BINARY=(pwd)/target/debug/hab-launch hab sup run
DEBUG: setting custom filesystem root for testing only (TESTING_FS_ROOT='/tmp/tmp.4aM77rLXOEhab-sup(MN): Enabling feature: TestExit
DEBUG: setting custom filesystem root for testing only (TESTING_FS_ROOT='/tmp/tmp.4aM77rLXOE')
hab-sup(MR): Supervisor Member-ID 8da7aa59e93c4215b9406d8610aac2a3
hab-sup(MR): Starting gossip-listener on 0.0.0.0:9638
hab-sup(MR): Starting ctl-gateway on 127.0.0.1:9632
hab-sup(MR): Starting http-gateway on 0.0.0.0:9631
```
In other shell:
```
➤ pgrep -afl  hab-
118443 /home/jbauman/habitat/target/debug/hab-launch run
118445 /home/jbauman/habitat/target/debug/hab-sup run

➤ echo "101" > /tmp/exit-file
```
```
hab-sup(MR): Simulating abrupt, unexpected exit with code 101
```
```
➤ pgrep -afl  hab-
118443 /home/jbauman/habitat/target/debug/hab-launch run

➤ pgrep -afl  hab-
118443 /home/jbauman/habitat/target/debug/hab-launch run
118661 /home/jbauman/habitat/target/debug/hab-sup run
```
```
hab-sup(MN): Enabling feature: TestExit
DEBUG: setting custom filesystem root for testing only (TESTING_FS_ROOT='/tmp/tmp.4aM77rLXOE')
hab-sup(MR): Supervisor Member-ID 8da7aa59e93c4215b9406d8610aac2a3
hab-sup(MR): Starting gossip-listener on 0.0.0.0:9638
hab-sup(MR): Starting ctl-gateway on 127.0.0.1:9632
hab-sup(MR): Starting http-gateway on 0.0.0.0:9631
```
That said, there's no _harm_ in calling `kill/wait` once `supervisor: Child` has already exited, so I can add that to the `drop` implementation for robustness.

Reply:
The more I think about it, you have a really good point. While I don't think it's broken the way the code is structured now, it's certainly not as robust as it should be.

Also, just to note, in the update I made the [`wait`](http://doc.rust-lang.org/1.28.0/std/process/struct.Child.html#method.wait) into a [`try_wait`](http://doc.rust-lang.org/1.28.0/std/process/struct.Child.html#method.try_wait). I don't think the latter existed at the time this code was originally written, but we absolutely don't want the `drop` function to block.